### ISSUE TITLE
[feature request] Improve multinomial sampling efficiency

### ISSUE THREAD

**Issue by Twice22** [tensorflow/tensorflow#19812]
### System information
- **Have I written custom code (as opposed to using a stock example script provided in TensorFlow)**: Yes
- **OS Platform and Distribution (e.g., Linux Ubuntu 16.04)**: Windows 10
- **TensorFlow installed from (source or binary)**: binary
- **TensorFlow version (use command below)**: b'v1.8.0-0-g93bc2e2072'
- **Python version**:  3.5.3
- **Bazel version (if compiling from source)**:
- **GCC/Compiler version (if compiling from source)**:
- **CUDA/cuDNN version**: 9.0
- **GPU model and memory**:  Quadro M1200
- **Exact command to reproduce**: 

### Describe the problem
I need to perform several samplings from a multinomial distribution. The problem is that it is very slow. After debugging, using a **Profiler**, I've realized that it comes from this piece of code:

`math_ops.reduce_sum(array_ops.one_hot(x, depth=k), axis=-2)` (line 257 at this moment)

from [this python file](https://github.com/tensorflow/tensorflow/blob/master/tensorflow/python/ops/distributions/multinomial.py)

Indeed, the code actually allocate a matrix for each row we want to sample. So it allocate too much memory for no actual reason! Here is my [stackoverflow post](https://stackoverflow.com/questions/50704004/tensorflow-efficient-multinomial-sampling-theano-x50-faster/50723793#50723793) for more detail about the problem

### Source code / logs
Actually Theano's implementation run x25 faster as described in my stackoverflow post. Also my (not so generic and not so fast solution) runs x3 times faster than the native TensorFlow sample function. Here is a snippet for comparison:

Using native implementation:
```python
import tensorflow as tf
import numpy as np
import tensorflow.contrib.distributions as ds
import time

tf.reset_default_graph()

nb_distribution = 100 # number of probabilities distribution

u = np.random.randint(2000, 3500, size=nb_distribution) # define number of counts (vector of size 100 with int in 2000, 3500)

# probsn is a matrix of probability:
# each row of probsn contains a vector of size 30 that sums to 1
probsn = np.random.uniform(size=(nb_distribution, 30))
probsn /= np.sum(probsn, axis=1)[:, None]

counts = tf.Variable(u, dtype=tf.float32)
probs = tf.Variable(tf.convert_to_tensor(probsn.astype(np.float32)))

# sample from the multinomial
dist = ds.Multinomial(total_count=counts, probs=probs)
out = dist.sample()


with tf.Session() as sess:
    sess.run(tf.global_variables_initializer())
    res = sess.run(out) # if remove this line the code is slower...
    start = time.time()
    res = sess.run(out)
    print(time.time() - start)
    print(np.all(u == np.sum(res, axis=1)))
```

On my computer it runs in **0.05** seconds

And here is my own (not generic) implementation of multinomial sampling that uses
`tf.scatter_nd()` function:

Using my own multinomial sampling:
```python
def vmultinomial_sampling(counts, pvals, seed=None):
    k = tf.shape(pvals)[1]
    logits = tf.expand_dims(tf.log(pvals), 1)

    def sample_single(args):
        logits_, n_draw_ = args[0], args[1]
        x = tf.multinomial(logits_, n_draw_, seed)
        indices = tf.cast(tf.reshape(x, [-1,1]), tf.int32)
        updates = tf.ones(n_draw_) # tf.shape(indices)[0]
        return tf.scatter_nd(indices, updates, [k])

    x = tf.map_fn(sample_single, [logits, counts], dtype=tf.float32)

    return x

xx = vmultinomial_sampling(u, probsn)
# check = tf.expand_dims(counts, 1) * probs

with tf.Session() as sess:
    sess.run(tf.global_variables_initializer())
    res = sess.run(xx) # if remove this line the code is slower...
    start_t = time.time()
    res = sess.run(xx)
    print(time.time() -start_t)
    #print(np.sum(res, axis=1))
    print(np.all(u == np.sum(res, axis=1)))
```

On my computer the code took **0.016** seconds to execute.

For comparison Theano's implementation takes **0.0025** seconds...

Indeed, my implementation doesn't take advantage of parallelization (changing the `parallel_iterations` in `map_fn` doesn't change anything. I'm using my GPU) while it totally makes sense that using parallel computations will improve the code because sampling from a specific row is independent from sampling from any other rows...

Do you think you can improve the code in order to avoid useless memory usage and allow parallel samplings?

Thank you.

**Comment by karmel** [comment]
@jvdillon - thoughts on the speed of sampling described by @Twice22 here?

**Comment by srvasude** [comment]
Hi!
  This can certainly be sped up to be much faster. The main issue is that the multinomial sampler uses a categorical sampler (unfortunately named tf.multinomial) underneath. We can actually construct an efficient binomial sampler, using a mix of inversion methods and Hormann's: https://www.tandfonline.com/doi/abs/10.1080/00949659308811496. From there, one can sample from a multinomial distribution through repeated binomial random variates.

**Comment by Twice22** [comment]
Hey.

I didn't look too much into the paper you mentioned, but it seems that, if I'm not mistaken, that Theano implementation is based on the "naive" idea. See [perform](https://github.com/Theano/Theano/blob/master/theano/sandbox/multinomial.py) function at line 155. I didn't take the time to look at the GPU implementation [here](https://github.com/Theano/Theano/blob/master/theano/gpuarray/multinomial.py).

Here is some benchmarks I did using numpy, C++ and I've also tried CUDA but didn't manage to fully implement it:

All the Benchmarks are more or less based on the same set-up (the random numbers associated to the 
counts are sampled in **[2000, 3500]**. For this reason the numbers to sample might vary from one experiment to another one but on average the time remains the same).

My CPU: Intel(R) Xeon(R) E3-1505M v6 @ 3.00Ghz
My GPU (Cuda compatible): Quadro M1200, Driver: 390.65

- Theano: 0.0025s
- TensorFlow [actual code]: 0.045s
- TensorFlow [my version]: 0.015s
- Numpy code: 0.62s
- C++ [Compile with Ox]: 0.25s
- CUDA  [couldn't make it work]

**Numpy version:**

```python
import numpy as np
import time

def vmultinomial(counts, pvals):
    N, K = pvals.shape
    Z = np.zeros_like(pvals, dtype='int32')
    cumsum = np.cumsum(pvals, axis=1, dtype='float64')
    
    for n in range(N):
        size_ = counts[n]
        rnd_numbers = np.random.uniform(size=size_)
        for i in range(counts[n]):
            Z[n, np.searchsorted(cumsum[n], rnd_numbers[i])] += 1
    
    return Z

nb_distribution = 100

u = np.random.randint(2000, 3500, size=nb_distribution)
probsn = np.random.uniform(size=(nb_distribution, 30))
probsn /= np.sum(probsn, axis=1)[:, None]

start_t = time.time()
res = vmultinomial(u, probsn)
print(time.time() - start_t)
```

**C++ code: **

```cpp
#include "stdafx.h"
#include <iostream>
#include <Eigen/Dense>
#include <algorithm>
#include <iostream>
#include <random>
#include <vector>
#include <chrono>

using namespace std;
using Eigen::MatrixXd;
using namespace std::chrono;


static vector<int> random_uniform(size_t size, int inf, int sup) {
	static uniform_int_distribution<int> distribution(inf, sup);
	static default_random_engine generator;

	// allocate a vector of size size
	vector<int> data(size);
	generate(data.begin(), data.end(), []() { return distribution(generator); });

	return data;
}

static vector<double> double_random_uniform(size_t size) {
	static uniform_real_distribution<double> distribution(0.0, 1.0);
	static default_random_engine generator;

	// allocate a vector of size size
	vector<double> data(size);
	generate(data.begin(), data.end(), []() { return distribution(generator); });

	return data;
}

MatrixXd probs_mat(size_t rows, size_t cols) {
	MatrixXd m = (MatrixXd::Random(rows, cols) + MatrixXd::Constant(rows, cols, 1.)) / 2;
	MatrixXd m2 = m.rowwise().sum();

	for (int i = 0; i < rows; i++) {
		for (int j = 0; j < cols; j++) {
			m(i, j) /= m2(i);
		}
	}
	
	return m;
}

MatrixXd cumsum(MatrixXd pvals) {
	int N = pvals.rows(), K = pvals.cols();
	MatrixXd Z(N, K);

	#pragma omp parallel for
	for (int i = 0; i < N; i++) {
		double sum = 0;
		for (int j = 0; j < K; j++) {
			sum += pvals(i, j);
			Z(i, j) = sum;
		}
	}

	return Z;
}

int searchsorted(MatrixXd csum, double rnd) {
	int beg = 0, end = csum.size();

	while (beg <= end) {
		int mid = (beg + end) / 2;
		if (csum(mid) < rnd) {
			beg = mid + 1;
		}
		else {
			end = mid - 1;
		}
	}
	return beg;
}


MatrixXd multinomial(vector<int> counts, MatrixXd probs) {
	int N = probs.rows(), K = probs.cols();
	MatrixXd Z = MatrixXd::Zero(N, K);

	MatrixXd csum = cumsum(probs);
	
#pragma omp parallel for
	for (int i = 0; i < N; i++) {
		int length = counts[i];
		vector<double> rnd_numbers = double_random_uniform(length);

		for (int j = 0; j < length; j++) {
			Z(i, searchsorted(csum.row(i), rnd_numbers[j])) += 1;
		}
	}
	return Z;
}

int main()
{
	int nb_samples = 100, pvals = 30;
	int low = 2000, high = 3500;
	vector<int> counts = random_uniform(nb_samples, low, high);
	MatrixXd probs = probs_mat(nb_samples, pvals);
	

	high_resolution_clock::time_point t1 = high_resolution_clock::now();
	MatrixXd csum = multinomial(counts, probs);
	high_resolution_clock::time_point t2 = high_resolution_clock::now();

	auto duration = duration_cast<microseconds>(t2 - t1).count();

	cout << duration << endl;
	system("pause");
}
```

**CUDA attempt:**
```cuda
#include <iostream>
#include <math.h>
#include "cuda_runtime.h"
#include "device_launch_parameters.h"
#include <stdio.h>
#include <ctime>

#include <thrust/device_vector.h>
#include <thrust/transform.h>
#include <thrust/iterator/counting_iterator.h>
#include <thrust/random.h>

#include <cublas_v2.h>
#include <curand.h>
#include <thrust/functional.h>
#include <thrust/iterator/counting_iterator.h>
#include <thrust/iterator/constant_iterator.h>

using namespace std;
using namespace thrust;

struct prg
{
	int a, b;

	__host__ __device__
		prg(int _a = 0, int _b = 1) : a(_a), b(_b) {};

	__host__ __device__
		int operator()(const unsigned int n) const
	{
		default_random_engine rng;
		uniform_int_distribution<int> dist(a, b);
		rng.discard(n);

		return dist(rng);
	}
};

struct float_prg
{
	int a, b;

	__host__ __device__
		float_prg(float _a = 0.0, float _b = 1.0) : a(_a), b(_b) {};

	__host__ __device__
		float operator()(const unsigned int n) const
	{
		default_random_engine rng;
		uniform_real_distribution<float> dist(a, b);
		rng.discard(n);

		return dist(rng);
	}
};


static device_vector<int> random_uniform(size_t size, int inf, int sup) {

	// allocate a vector of size size
	device_vector<int> data(size);
	counting_iterator<unsigned int> index_sequence_begin(0);

	transform(index_sequence_begin, index_sequence_begin + size, data.begin(), prg(inf, sup));

	return data;
}

__host__ __device__
static device_vector<float> double_random_uniform(size_t size) {
	// allocate a vector of size size
	device_vector<float> data(size);
	counting_iterator<unsigned int> index_sequence_begin(0);

	transform(index_sequence_begin, index_sequence_begin + size, data.begin(), float_prg(0.0, 1.0));

	return data;
}

// don't care if this function doesn't use CUDA, I don't eval
// its performance
void normalize_l1(device_vector<float> &A, int rows, int cols) {
	float *row_sum = (float *)malloc(rows * sizeof(float));

	for (int i = 0; i < rows; ++i) {
		float sum = 0.0f;
		for (int j = 0; j < cols; ++j) {
			sum += A[cols*i + j];
		}
		row_sum[i] = sum;
	}

	for (int r = 0; r < rows; ++r) {
		for (int c = 0; c < cols; ++c) {
			A[r*cols + c] /= row_sum[r];
		}
	}
}

void GPU_fill_rand(device_vector<float> &A, int nr_rows_A, int nr_cols_A) {
	// Create a pseudo-random number generator
	curandGenerator_t prng;
	curandCreateGenerator(&prng, CURAND_RNG_PSEUDO_DEFAULT);

	// Set the seed for the random number generator using the system clock
	curandSetPseudoRandomGeneratorSeed(prng, (unsigned long long) clock());

	// Fill the array with random numbers on the device
	curandGenerateUniform(prng, raw_pointer_cast(&A[0]), nr_rows_A * nr_cols_A);

	// normalize the matrix
	normalize_l1(A, nr_rows_A, nr_cols_A);

}

template<class T>
void print_matrix(const device_vector<T> &A, int nr_rows_A, int nr_cols_A) {

	for (int i = 0; i < nr_rows_A; ++i) {
		for (int j = 0; j < nr_cols_A; ++j) {
			cout << A[i * nr_cols_A + j] << " ";
		}
		cout << endl;
	}
	cout << endl;
}

__device__
int searchsorted(float* csum, int beg, int end, double rnd) {

	while (beg <= end) {
		int mid = (beg + end) / 2;
		if (csum[mid] < rnd) {
			beg = mid + 1;
		}
		else {
			end = mid - 1;
		}
	}
	return beg;
}

// compute the cumulative sum over the row of the input matrix Z
void cumsum(device_vector<float> &Z, int N, int K) {

	device_vector<float> cumProb(N*K, 0.);
	device_vector<float> sub(N*K);

	inclusive_scan(Z.begin(), Z.end(), Z.begin());

	transform(make_counting_iterator(0), make_counting_iterator(N*K), make_constant_iterator(K), sub.begin(), thrust::divides<int>());

	transform(Z.begin(), Z.end(), sub.begin(), Z.begin(), thrust::minus<float>());
}

// Try to transform kernel into global function. Failed
__global__ void kernel(int* Z, int* counts, float* ptr_probs, int N, int K) {
	int idx = threadIdx.x + blockIdx.x * blockDim.x;
	int stride = blockDim.x * gridDim.x;

	for (int i = idx; i < N; i += stride) {
		int length = counts[i];
		device_vector<float> rnd_numbers = double_random_uniform(length);
		float* ptr_rnd_numbers = raw_pointer_cast(rnd_numbers.data());

		for (int j = 0; j < length; j++) {
			Z[searchsorted(ptr_probs, i*K, (i + 1)*K, ptr_rnd_numbers[j])] += 1;
		}
	}
}

device_vector<int> multinomial(device_vector<int> &counts, device_vector<float> probs, int N, int K) {
	// probs -> cumulative sum over rows
	cumsum(probs, N, K);
	device_vector<int> Z(N*K, 0);

	int* ptr_counts = raw_pointer_cast(counts.data());
	float* ptr_probs = raw_pointer_cast(probs.data());

	kernel<<< 32, 64 >>>(raw_pointer_cast(Z.data()), ptr_counts, ptr_probs, N, K);

	return Z;
}

int main(void)
{

	int nb_samples = 3, pvals = 4;
	int low = 2000, high = 3500;
	
	device_vector<int> counts = random_uniform(nb_samples, low, high);

	// Allocate 1 matrix on GPU
	device_vector<float> d_probs(nb_samples * pvals);

	// fill d_probs with random number s.t sum(rows) = 1
	GPU_fill_rand(d_probs, nb_samples, pvals);

	device_vector<int> res = multinomial(counts, d_probs, nb_samples, pvals);

	print_matrix(res, nb_samples, pvals);

	return 0;
}
```

The idea for the CUDA implementation would be to use multiple threads to parallelize the double loop in the multinomial function. I think that should speed up the process quite a bit? I think having a x100 speed-up as compare to the simple CPU C++ implementation is reachable ?

**Comment by tensorflowbutler** [comment]
Nagging Assignee @srvasude: It has been 15 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
Nagging Assignee @srvasude: It has been 32 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
Nagging Assignee @srvasude: It has been 47 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
Nagging Assignee @srvasude: It has been 62 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
Nagging Assignee @srvasude: It has been 77 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
Nagging Assignee @srvasude: It has been 92 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
Nagging Assignee @srvasude: It has been 107 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
Nagging Assignee @srvasude: It has been 137 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
Nagging Assignee @srvasude: It has been 152 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
Nagging Assignee @srvasude: It has been 167 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by srvasude** [comment]
Closing this issue.

TF distributions are deprecated / removed. We've had a similar issue in TensorFlow Probability for multinomial. The implementation there is based on a map_fn as well. We can still do better by using the random_binomial sampler in TF (which is rejection based, and will probably work much better in the limit of large counts). Feel free to reach out at https://github.com/tensorflow/probability.
